File: parsers/transfer_list.py
"""
Transfer List Excel Parser
Parses City of Saskatoon property transfer list (Excel/XLSX).
Columns: Roll #, Civic_Address, Vendor, Purchaser, Sales_Date, Sales_Price, PPT, PPT_Descriptor
"""

from typing import Optional, List
import openpyxl
import json
import sys
from datetime import datetime
from pathlib import Path


def parse_transfer_list(xlsx_path: str) -> List[dict]:
    """Parse transfer list Excel file into structured records."""
    wb = openpyxl.load_workbook(xlsx_path, read_only=True)
    ws = wb.active

    records = []
    headers = None

    for row in ws.iter_rows(values_only=True):
        # First row is headers
        if headers is None:
            # Take only meaningful columns (first 8)
            headers = [str(h).strip() if h else f"col_{i}" for i, h in enumerate(row[:8])]
            continue

        # Skip empty rows
        vals = row[:8]
        if not any(vals):
            continue

        roll_number = str(vals[0]) if vals[0] else None
        address = str(vals[1]).strip() if vals[1] else None
        vendor = _clean_name(vals[2])
        purchaser = _clean_name(vals[3])
        sales_date = _format_date(vals[4])
        sales_price = float(vals[5]) if vals[5] else None
        ppt_code = str(vals[6]).strip() if vals[6] else None
        ppt_descriptor = str(vals[7]).strip() if vals[7] else None

        records.append({
            "roll_number": roll_number,
            "address": address,
            "vendor": vendor,
            "purchaser": purchaser,
            "sales_date": sales_date,
            "sales_price": sales_price,
            "property_type_code": ppt_code,
            "property_type": ppt_descriptor,
        })

    wb.close()
    return records


def _clean_name(val) -> Optional[str]:
    """Clean up a company/person name (remove newlines, extra whitespace)."""
    if not val:
        return None
    return " ".join(str(val).split())


def _format_date(val) -> Optional[str]:
    """Format date to ISO string."""
    if not val:
        return None
    if isinstance(val, datetime):
        return val.strftime("%Y-%m-%d")
    return str(val)


def main():
    if len(sys.argv) < 2:
        print("Usage: python transfer_list.py <xlsx_path> [output.json]")
        sys.exit(1)

    xlsx_path = sys.argv[1]
    records = parse_transfer_list(xlsx_path)

    print(f"Parsed {len(records)} transfer records")

    if len(sys.argv) > 2:
        output_path = sys.argv[2]
        Path(output_path).write_text(json.dumps(records, indent=2))
        print(f"Written to {output_path}")
    else:
        # Print summary
        for r in records[:5]:
            print(json.dumps(r, indent=2))
        if len(records) > 5:
            print(f"... and {len(records) - 5} more")


if __name__ == "__main__":
    main()

File: src/lib/parsers/transfer-list.ts
/**
 * Transfer List Excel Parser (TypeScript)
 * Parses City of Saskatoon property transfer list (XLSX).
 */

import * as XLSX from "xlsx";

export interface TransferRecord {
  rollNumber: string | null;
  address: string | null;
  vendor: string | null;
  purchaser: string | null;
  salesDate: string | null;
  salesPrice: number | null;
  propertyTypeCode: string | null;
  propertyType: string | null;
}

function cleanName(val: unknown): string | null {
  if (!val) return null;
  return String(val).replace(/\r?\n/g, " ").replace(/\s+/g, " ").trim();
}

function formatDate(val: unknown): string | null {
  if (!val) return null;
  if (typeof val === "number") {
    // Excel serial date number
    const date = XLSX.SSF.parse_date_code(val);
    if (date) {
      return `${date.y}-${String(date.m).padStart(2, "0")}-${String(date.d).padStart(2, "0")}`;
    }
  }
  if (val instanceof Date) {
    return val.toISOString().split("T")[0];
  }
  return String(val);
}

export function parseTransferList(buffer: Buffer | Uint8Array): TransferRecord[] {
  const wb = XLSX.read(buffer, { type: "buffer" });
  const ws = wb.Sheets[wb.SheetNames[0]];
  const rows: unknown[][] = XLSX.utils.sheet_to_json(ws, { header: 1 });

  const records: TransferRecord[] = [];
  // First row is headers, skip it
  for (let i = 1; i < rows.length; i++) {
    const row = rows[i];
    if (!row || !row.length || !row.some((v) => v != null && v !== "")) continue;

    records.push({
      rollNumber: row[0] != null ? String(row[0]) : null,
      address: row[1] != null ? String(row[1]).trim() : null,
      vendor: cleanName(row[2]),
      purchaser: cleanName(row[3]),
      salesDate: formatDate(row[4]),
      salesPrice: row[5] != null ? Number(row[5]) : null,
      propertyTypeCode: row[6] != null ? String(row[6]).trim() : null,
      propertyType: row[7] != null ? String(row[7]).trim() : null,
    });
  }

  return records;
}
